tests/test_compiler.py

```diff
@@ -405,6 +423,57 @@ def inner(j):
 
         assert jnp.allclose(circuit(4), jnp.eye(2**4)[0])
 
+    def test_for_loop_python_fallback(self):
+        """Test that qml.for_loop fallsback to Python
+        interpretation if Catalyst is not available"""
+        dev = qml.device("lightning.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit(x, n):
+
+            # for loop with dynamic bounds
+            @qml.for_loop(0, n, 1)
+            def loop_fn(i):
+                qml.Hadamard(wires=i)
+
+            # nested for loops.
+            # outer for loop updates x
+            @qml.for_loop(0, n, 1)
+            def loop_fn_returns(i, x):
+                qml.RX(x, wires=i)
+
+                # inner for loop
+                @qml.for_loop(i + 1, n, 1)
+                def inner(j):
+                    qml.CRY(x**2, [i, j])
+
+                inner()
+
+                return jnp.sin(x)
+
+            loop_fn()
+            loop_fn_returns(x)
+
+            return qml.expval(qml.PauliZ(0))
+
+        x = 0.5
+        assert jnp.allclose(circuit(x, 2), qml.qjit(circuit)(x, 2))
+
+        res = circuit.tape.operations
+        expected = [
+            qml.Hadamard(wires=[0]),
+            qml.Hadamard(wires=[1]),
+            qml.Hadamard(wires=[2]),
+            qml.RX(0.5, wires=[0]),
+            qml.CRY(0.25, wires=[0, 1]),
+            qml.CRY(0.25, wires=[0, 2]),
+            qml.RX(0.6, wires=[1]),
+            qml.CRY(0.36, wires=[1, 2]),
+            qml.RX(0.7, wires=[2]),
+        ]
+
+        assert [qml.equal(i, j) for i, j in zip(res, expected)]
```

Comment:
```suggestion
        _ = [qml.assert_equal(i, j) for i, j in zip(res, expected)]
```

If any of them fail, the error message will contain information about why it failed.  This will make test debugging easier going forward.

Reply:
nice!